### PR TITLE
Neo4j subgraph isomorphism

### PR DESCRIPTION
**Pull request by jiatistuta**
The concurrent version of the plugin is ready for review. 
Two parameters have been added to the procedure: 
parallelFactor: the number of threads used to do the job is defined by parallelFacto r* numOfCPUcores
suppressResult: True-in case of too many nodes increasing display overhead, only display number of subgraphs and execution time.
                          False-display full results

example: `call subgraphIso("Query","Target","2","True")

### CODE REVIEW

**Review by scordio — CHANGES_REQUESTED** [review]
Please apply these changes while I do the other tests:
 * All code should not contain any IDE related file, therefore please remove the `.idea` folder, the `Subgraph_Isomorphism.iml` file and the `target` folder
* All classes should follow the CamelCase naming convention, i.e., class names should be nouns, in mixed case with the first letter of each internal word capitalized, no other symbols are allowed

Thanks

File: neo4j-plugins/subgraph-isomorphism/pom.xml [hunk 1]
```diff
@@ -0,0 +1,103 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<project xmlns="http://maven.apache.org/POM/4.0.0"
+         xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+    <modelVersion>4.0.0</modelVersion>
+
+    <groupId>edu.msstate.dasi</groupId>
```
Comment (scordio) [inline neo4j-plugins/subgraph-isomorphism/pom.xml]: Change to `<groupId>edu.msstate.dasi.csb.neo4j</groupId>`

File: neo4j-plugins/subgraph-isomorphism/pom.xml [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+    <artifactId>subgraph-isomorphism</artifactId>
+    <version>0.1-SNAPSHOT</version>
+
+    <packaging>jar</packaging>
+    <name>Subgraph Isomorphism</name>
+    <description>A subgraph isomorphism algorithm for the Neo4j graph database</description>
+
+    <properties>
+        <neo4j.version>3.1.4</neo4j.version>
```
Comment (scordio) [inline neo4j-plugins/subgraph-isomorphism/pom.xml]: Update Neo4j version to 3.2.0

File: neo4j-plugins/subgraph-isomorphism/pom.xml [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+    </properties>
```
Comment (scordio) [inline neo4j-plugins/subgraph-isomorphism/pom.xml]: Add new property: `<neo4j-driver.version>1.3.0</neo4j-driver.version>`

File: neo4j-plugins/subgraph-isomorphism/pom.xml [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+            <artifactId>neo4j-harness</artifactId>
+            <version>${neo4j.version}</version>
+            <scope>test</scope>
+        </dependency>
+
+        <dependency>
+            <!-- Used to send cypher statements to our procedure. -->
+            <groupId>org.neo4j.driver</groupId>
+            <artifactId>neo4j-java-driver</artifactId>
+            <version>1.2.0</version>
```
Comment (scordio) [inline neo4j-plugins/subgraph-isomorphism/pom.xml]: Replace with `<version>${neo4j-driver.version}</version>`

File: neo4j-plugins/subgraph-isomorphism/pom.xml [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+                <executions>
+                    <execution>
+                        <phase>package</phase>
+                        <goals>
+                            <goal>shade</goal>
+                        </goals>
+                    </execution>
+                </executions>
+            </plugin>
+            <plugin>
```
Comment (scordio) [inline neo4j-plugins/subgraph-isomorphism/pom.xml]: Remove the `maven-jar-plugin` section

File: neo4j-plugins/subgraph-isomorphism/src/main/java/META-INF/MANIFEST.MF [hunk 6]
```diff
@@ -0,0 +1,3 @@
+Manifest-Version: 1.0
```
Comment (scordio) [inline neo4j-plugins/subgraph-isomorphism/src/main/java/META-INF/MANIFEST.MF]: Remove the whole file and the `META-INF` directory

**Review by scordio — CHANGES_REQUESTED** [review]

File: neo4j-plugins/subgraph-isomorphism/src/main/java/META-INF/MANIFEST.MF (hunk 6 above)
Comment (jiatistuta) [reply neo4j-plugins/subgraph-isomorphism/src/main/java/META-INF/MANIFEST.MF]: All changes are done. Thank you.

**Comment by scordio** [comment]
Is it possible to simplify the output?

Right now it is:

````
╒═════════════╤════════════╤═══════════════╤══════════════════╤═══════════════╕
│"resultNode" │"queryNode" │"subgraphIndex"│"totalNumSubgraph"│"executionTime"│
╞═════════════╪════════════╪═══════════════╪══════════════════╪═══════════════╡
│{"name":"11"}│{"name":"1"}│"0"            │"2"               │"85ms"         │
├─────────────┼────────────┼───────────────┼──────────────────┼───────────────┤
│{"name":"12"}│{"name":"2"}│"0"            │"2"               │"85ms"         │
├─────────────┼────────────┼───────────────┼──────────────────┼───────────────┤
│{"name":"13"}│{"name":"3"}│"0"            │"2"               │"85ms"         │
├─────────────┼────────────┼───────────────┼──────────────────┼───────────────┤
│{"name":"11"}│{"name":"1"}│"1"            │"2"               │"85ms"         │
├─────────────┼────────────┼───────────────┼──────────────────┼───────────────┤
│{"name":"13"}│{"name":"2"}│"1"            │"2"               │"85ms"         │
├─────────────┼────────────┼───────────────┼──────────────────┼───────────────┤
│{"name":"14"}│{"name":"3"}│"1"            │"2"               │"85ms"         │
└─────────────┴────────────┴───────────────┴──────────────────┴───────────────┘
````
with `suppressResult` as `false`, or:

````
╒════════════╤═══════════╤═══════════════╤══════════════════╤═══════════════╕
│"resultNode"│"queryNode"│"subgraphIndex"│"totalNumSubgraph"│"executionTime"│
╞════════════╪═══════════╪═══════════════╪══════════════════╪═══════════════╡
│null        │null       │null           │"2"               │"10ms"         │
└────────────┴───────────┴───────────────┴──────────────────┴───────────────┘
````
with `suppressResult` as `true`.

Basically, we don't want to count the number of subgraphs as it could be obtained from the result and we don't want to compute the execution time as it is provided by Neo4j natively.

Furthermore, right now giving only the total number of subgraphs is not a requirement, so there is no need to have the `suppressResult` switch. The computation for bigger graphs will take the time that it takes.

A simplified output could have this shape:
* One column for each pattern vertex, ~~with its pattern ID in the header (I would suggest to use the Neo4j ID in the header instead of the vertex name in order to avoid vertices with the same name)~~ (on second thought, using the vertex name is fine and more user-friendly)
* One row for each subgraph, showing the corresponding vertex in the bigger graph

See the following as an example:

````
╒═════════════╤═════════════╤═════════════╕
│"1"          │"2"           │"3"         │
╞═════════════╪═════════════╪═════════════╡
│{"name":"11"}│{"name":"13"}│{"name":"14"}│
├─────────────┼─────────────┼─────────────┤
│{"name":"11"}│{"name":"12"}│{"name":"13"}│
└─────────────┴─────────────┴─────────────┘
````
`1`, `2` and `3` are pattern vertices, while `11`, `12`, `13` and `14` are the matching vertices in the bigger graph.

I used the following for this test/example:
* Pattern: pattern
* Graph: graph

Thanks

**Comment by scordio** [comment]
Using the following both as a pattern and as a graph:

screen20

I receive the following output:

````
╒════════════╤════════════╤═══════════════╤══════════════════╤═══════════════╕
│"resultNode"│"queryNode" │"subgraphIndex"│"totalNumSubgraph"│"executionTime"│
╞════════════╪════════════╪═══════════════╪══════════════════╪═══════════════╡
│{"name":"8"}│{"name":"8"}│"0"            │"1"               │"74ms"         │
├────────────┼────────────┼───────────────┼──────────────────┼───────────────┤
│{"name":"1"}│{"name":"1"}│"0"            │"1"               │"74ms"         │
├────────────┼────────────┼───────────────┼──────────────────┼───────────────┤
│{"name":"2"}│{"name":"2"}│"0"            │"1"               │"74ms"         │
├────────────┼────────────┼───────────────┼──────────────────┼───────────────┤
│{"name":"3"}│{"name":"3"}│"0"            │"1"               │"74ms"         │
├────────────┼────────────┼───────────────┼──────────────────┼───────────────┤
│{"name":"4"}│{"name":"4"}│"0"            │"1"               │"74ms"         │
├────────────┼────────────┼───────────────┼──────────────────┼───────────────┤
│{"name":"5"}│{"name":"5"}│"0"            │"1"               │"74ms"         │
├────────────┼────────────┼───────────────┼──────────────────┼───────────────┤
│{"name":"6"}│{"name":"6"}│"0"            │"1"               │"74ms"         │
├────────────┼────────────┼───────────────┼──────────────────┼───────────────┤
│{"name":"7"}│{"name":"7"}│"0"            │"1"               │"74ms"         │
└────────────┴────────────┴───────────────┴──────────────────┴───────────────┘
````
which corresponds to a single subgraph, but it should return also the other possibilities.

In this particular case there are a lot of subgraphs, more than 5000 (you can try with a smaller graph which has similar concentric structure).

Thanks

**Comment by jiatistuta** [comment]
Stefano,

I tried and current framework doesn’t support this output format. 

The output to the console is a Stream<Result> where Result is the customized class defining the content of the output. Each instance variable represents a column. Multiple objects in the Stream create multiple rows along the columns. Here is the class in our plugin:

public class Result {

    public Node resultNode;
    public Node queryNode;
    public String subgraphIndex;
    public String totalNumSubgraph;
    public String executionTime;
    public String[] NodeIDList;

    public Result(Node resultNode, Node queryNode, String subgraphIndex, String totalNumSubgraph, String executionTime, String [] NodeIDList) {
        this.resultNode = resultNode;
        this.queryNode = queryNode;
        this.subgraphIndex = subgraphIndex;
        this.totalNumSubgraph=totalNumSubgraph;
        this.executionTime=executionTime;
        this.NodeIDList=NodeIDList;
    }
}


Therefore, if we want the columns to be the patter vertexes, we need to dynamically change the number of the member variables in that class to the number of vertexes in the pattern/query graph.

Maybe output this simplified form to an external file instead of in the console? Any thoughts?

Thank you for the comment.

**Comment by jiatistuta** [comment]
Yes, I did this on purpose. In this case, there should be all possibilities in the permutation returned. I removed all other subgraphs in the permutation but leave only one subgraph. In my opinion they have the same topological structure and may induce redundant information.

I can fix this to show all possibilities.

**Comment by scordio** [comment]
Let me think about the output issue, I'll let you know ASAP.

Regarding the permutations, we can't make assumptions about structural redundancy in advance and we should return all the subgraphs.

Thanks

**Comment by jiatistuta** [comment]
Now all possibilities will be displayed. Unique node ID is displayed instead of other properties. 

Sent from Mail for Windows 10

**Comment by jiatistuta** [comment]
> (on second thought, using the vertex name is fine and more user-friendly)

Sure, it is OK to use "name" for display. However, the vertex name is one of the properties of a node. If we explicitly use the "name" property, is it a problem when a user wants to show something else? For example, maybe in a graph there are no "name" property attached to any nodes.

By default, if we put a Node object to display, it will display all properties of the node (the previous version of the plugin). I think we can let the user choose which property to display. Add a parameter to the procedure can do the work. 

I have already added the parameter "selectProperty". 

Usage is [here](https://github.com/msstate-dasi/csb/blob/neo4j-subgraph-isomorphism/neo4j-plugins/subgraph-isomorphism/README.md)


Please let me know your comments.

**Comment by scordio** [comment]
# Input
We should offer a parameter set that is more user-oriented.
* _queryLabel_: OK but I would rename it to _patternLabel_ to avoid confusion with Cypher queries.
* _targetLabel_: OK but don't use "All" as a special case because it is an allowed label, instead you could use the empty string.
* ~~_The parallelism factor & the split size_~~: these parameters are specific to the internal implementation of the procedure and I don't see the need to expose these details to the end user since it's not a distributed computing platform. What I would expect from the procedure is to maximize the resource consumption on any machine (e.g. use as many thread as possible), is it possible to achieve? Can you simply remove the `parallelism` parameter of the `ForkJoinPool` constructor? The basic idea is that the user should not worry about specifying the level of parallelism but it's the procedure that should do always its best according to the environment, and the work stealing feature of the pool seems enough to me.
* ~~_selectProperty_~~: my suggestion about the ID/name was to be considered only if we could define a dynamic header, therefore we can return the whole Node (see also the next section) and there is no need to keep this parameter.
* ~~_Suppressed result_~~: we want to display always the results, therefore there is no need to keep this parameter.

# Output
* Please change the `subgraphIndex` type from `String` to `Long`
* Please replace any `query` occurrencies with `pattern` as `query` can be confused with Cyper queries. 
* Please drop `totalNumSubgraph` as it can be computed using the original result:
````sql
CALL SubgraphIso("patternLabel", "graphLabel", "2", "100", "False") YIELD subgraphIndex
RETURN count(DISTINCT subgraphIndex)
````
* Since it is not possible to define a dynamic result format, we can return the whole `Node` object.

Altogether, `Result` fields names and order should be:
````java
public Long subgraphIndex;
public Node patternNode;
public Node targetNode;
````

# General considerations
* The package name in the Java source should be `edu.msstate.dasi.csb.neo4j` instead of `subgraphIso` 
* The procedure name and main method should be full and should follow the lowerCamelCase convention: `subgraphIsomorphism`
* The `UllmannAlg` method should follow the lowerCamelCase convention
* It would be good to add a prefix to the exposed procedure name with a package identifier (but not the whole package name) to avoid overlapping with other procedures, something like:
````java
@Procedure("csb.subgraphIsomorphism")
````
* The description annotation is not updated, it could be something like:
````java
@Description("CALL csb.subgraphIsomorphism(patternLabel, targetLabel) YIELD subgraphIndex, patternNode, targetNode  - finds subgraphs of the target graph which are isomorphic to the pattern")
````
* Please adapt the `@Name` annotations to have consistent names everywhere (usage, variables, etc.).

**Comment by jiatistuta** [comment]
> **What I would expect from the procedure is to maximize the resource consumption on any machine (e.g. use as many thread as possible), is it possible to achieve?**

I have done experiments with the same # of subgraphs and different number of threads. Surprisingly, increasing number of threads may degrade performance. I think this is related to the actual number of physical CPU cores and number of tasks. I don't think there is a one-for-all parameter configuration that can fit all problems. [This article](http://faculty.ycp.edu/~dhovemey/spring2013/cs365/lecture/lecture18.html) has detailed discussion about this topic. Like it said:

- >By using a thread pool with work stealing, a fork/join framework can allow a relatively fine-grained division of the problem, but only create the minimum number of threads needed to fully exploit the available CPU cores. Typically, the thread pool will have one thread per available CPU core.

- >One of the main things to consider when implementing an algorithm using fork/join parallelism is choosing the threshold which determines whether a task will execute a sequential computation rather than forking parallel sub-tasks. If the threshold is too large, then the program might not create enough tasks to fully take advantage of the available processors/cores. If the threshold is too small, then the overhead of task creation and management could become significant. In general, some experimentation will be necessary to find an appropriate threshold value.

According to my experiments, the best performance is achieved when the number of threads is twice the number of CPU cores and the # of tasks is greater than the number of threads but less than what the heap size can handle. More tasks don't show obvious performance gain, yet no loss.

> **Can you simply remove the parallelism parameter of the ForkJoinPool constructor?**

No. The default value of parallelism is `Runtime.availableProcessors()`, which is the number of logical processors in the computer. For example, in a hyper-threading 4-core CPU system `Runtime.availableProcessors()=8`

For now, a safe configuration is:
- Number of threads=`2*Runtime.availableProcessors()`
- Number of tasks=2*number of threads, or as close as possible if number of tasks is less.

> **Please change the subgraphIndex type from String to Long**

The Neo4j only accepts two types as member variables of the Result class: Node and String. If we define subgraphIndex as Long, you can't even open the database. In this case, String is converted from Long by `Long.toString()` so there should be no accuracy loss.

> **~~Suppressed result~~: we want to display always the results, therefore there is no need to keep this parameter.**

Yes I can do it. My only concern is, if there are too many results popped up, say, 362880, how could the user efficiently browse and check the results?

I will make all the changes quickly.

**Comment by scordio** [comment]
> No. The default value of parallelism is Runtime.availableProcessors(), which is the number of logical processors in the computer. For example, in a hyper-threading 4-core CPU system Runtime.availableProcessors()=8
>
> For now, a safe configuration is:
>
>    Number of threads=2*Runtime.availableProcessors()
>    Number of tasks=2*number of threads, or as close as possible if number of tasks is less.

Ok I see, then let's keep this parameter as a long parameter (see the next comment about the type) and you can also define a default value based on your analysis so that the procedure can be called omitting the `parallelFactor` parameter. There is an example [here](https://neo4j.com/docs/developer-manual/current/extending-neo4j/procedures/) on how to define a default value for procedure parameters, an annotation like the following should be good:
````java
@Name(value = "parallelFactor", defaultValue = theDefaultValue)
````
> The Neo4j only accepts two types as member variables of the Result class: Node and String. If we define subgraphIndex as Long, you can't even open the database. In this case, String is converted from Long by Long.toString() so there should be no accuracy loss.

I tried before writing the comment and it worked, also see [here](https://neo4j.com/docs/java-reference/current/javadocs/org/neo4j/procedure/Procedure.html).

> Yes I can do it. My only concern is, if there are too many results popped up, say, 362880, how could the user efficiently browse and check the results?

I tried through the console and the output was actually huge, but it is not our responsibility to tell the user how to handle this kind of output, it's like saying that the unix cat command should limit the output in case of huge data. As an example, the user could redirect, filter or limit the output of the procedure.

Thanks

**Comment by jiatistuta** [comment]
I have applied all your comments to the code. Codes have been pushed to the repository. Default value of `parallelFactor` is 2. The number of tasks is hard coded as `2*ForkJoinPool.getParallelism()`.

> I tried before writing the comment and it worked, also see here.

It looks like a bug in Neo4j 3.1.4. After updating my Neo4j to 3.2.0, it works fine.

**Review by scordio — APPROVED** [review]

**Pull request merged**